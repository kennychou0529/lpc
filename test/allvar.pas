                                (* Chapter 3 - Program 3 *)
program AllSimpleVariableTypes;

var  A,B         : integer;
     C,D         : 0..255;
     DogTail    : real;
     Puppy       : boolean;
     AnimalCookies : char;

begin
   A := 4;
   B := 5;
   C := 212;
   D := C + 3;
   DogTail := 345.12456;
   Puppy := B > A;  (* since B is greater than A, Puppy
                       will be assigned the value TRUE *)
   AnimalCookies := 'R';  (* this is a single character *)

   Writeln('The integers are',A:5,B:5);
   Writeln('The bytes are',   C:5,D:5); (* notice that the spaces
                                           prior to the C will be
                                           ignored on output *)
   Writeln('The real value is',DogTail:12:2,DogTail:12:4);
   Writeln;
   Writeln('The boolean value is ',Puppy,Puppy:13);
   Writeln('The char variable is an ',AnimalCookies);
end.




{ Result of execution

The integers are    4    5
The bytes are  212  215
The real value is      345.12    345.1246

The boolean value is TRUE         TRUE
The char variable is an R

}
