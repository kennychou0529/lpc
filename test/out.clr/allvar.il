
#include "commonDefs.il"

.field public static string[] _ProgramArgs

.method static void .cctor()
{
    ldc.i4 2
    newarr string
    dup
    ldc.i4 0
    ldstr "_input"
    stelem string
    dup
    ldc.i4 1
    ldstr "_output"
    stelem string
    stsfld string[] _ProgramArgs
    ret
}

// "allvar.pas"

// line 2
//================================================================================
// scope: program scope (level : 1)

// global variables
.class _G
{
    .field public static valuetype [lpcRuntime]LPC.File _input
    .field public static valuetype [lpcRuntime]LPC.File _output
    .field public static int32 A
    .field public static uint8 AnimalCookies
    .field public static int32 B
    .field public static uint8 C
    .field public static uint8 D
    .field public static float64 DogTail
    .field public static bool Puppy
}

// line 0
//================================================================================
// scope: global scope (level : 0)

// line 2

.method static
void _Program()
{
    .maxstack 32

    .locals init()

    ldsflda valuetype [lpcRuntime]LPC.File _G::_input
    ldc.i4 0
    call instance void [lpcRuntime]LPC.File::Init(int32)

    ldsflda valuetype [lpcRuntime]LPC.File _G::_output
    ldc.i4 1
    call instance void [lpcRuntime]LPC.File::Init(int32)

L_Entry:
    // line 11
    ldc.i4 4
    stsfld int32 _G::A
    // line 12
    ldc.i4 5
    stsfld int32 _G::B
    // line 13
    ldc.i4 212
    dup
    ldc.i4 0
    ldc.i4 255
    call void [lpcRuntime]LPC.Runtime::RTCheckRange(int32, int32, int32)
    stsfld uint8 _G::C
    // line 14
    ldsfld uint8 _G::C
    ldc.i4 3
    add
    dup
    ldc.i4 0
    ldc.i4 255
    call void [lpcRuntime]LPC.Runtime::RTCheckRange(int32, int32, int32)
    stsfld uint8 _G::D
    // line 15
    ldc.r8 345.125
    stsfld float64 _G::DogTail
    // line 16
    ldsfld int32 _G::B
    ldsfld int32 _G::A
    cgt
    stsfld bool _G::Puppy
    // line 18
    ldc.i4 82 // 'R'
    stsfld uint8 _G::AnimalCookies
    // line 20
    ldsflda valuetype [lpcRuntime]LPC.File _G::_output
    dup
    ldstr "The integers are"
    call instance void [lpcRuntime]LPC.File::Write(string)
    dup
    ldc.i4 5
    ldc.i4.0
    ldsfld int32 _G::A
    call instance void [lpcRuntime]LPC.File::Format(int32, int32, int32)
    dup
    ldc.i4 5
    ldc.i4.0
    ldsfld int32 _G::B
    call instance void [lpcRuntime]LPC.File::Format(int32, int32, int32)
    call instance void [lpcRuntime]LPC.File::WriteLn()
    // line 21
    ldsflda valuetype [lpcRuntime]LPC.File _G::_output
    dup
    ldstr "The bytes are"
    call instance void [lpcRuntime]LPC.File::Write(string)
    dup
    ldc.i4 5
    ldc.i4.0
    ldsfld uint8 _G::C
    call instance void [lpcRuntime]LPC.File::Format(int32, int32, int32)
    dup
    ldc.i4 5
    ldc.i4.0
    ldsfld uint8 _G::D
    call instance void [lpcRuntime]LPC.File::Format(int32, int32, int32)
    call instance void [lpcRuntime]LPC.File::WriteLn()
    // line 24
    ldsflda valuetype [lpcRuntime]LPC.File _G::_output
    dup
    ldstr "The real value is"
    call instance void [lpcRuntime]LPC.File::Write(string)
    dup
    ldc.i4 12
    ldc.i4 2
    ldsfld float64 _G::DogTail
    call instance void [lpcRuntime]LPC.File::Format(int32, int32, float64)
    dup
    ldc.i4 12
    ldc.i4 4
    ldsfld float64 _G::DogTail
    call instance void [lpcRuntime]LPC.File::Format(int32, int32, float64)
    call instance void [lpcRuntime]LPC.File::WriteLn()
    // line 25
    ldsflda valuetype [lpcRuntime]LPC.File _G::_output
    call instance void [lpcRuntime]LPC.File::WriteLn()
    // line 26
    ldsflda valuetype [lpcRuntime]LPC.File _G::_output
    dup
    ldstr "The boolean value is "
    call instance void [lpcRuntime]LPC.File::Write(string)
    dup
    ldsfld bool _G::Puppy
    brtrue L_bool_to_string_true_1
    ldstr "false"
    br L_bool_to_string_false_2
L_bool_to_string_true_1:
    ldstr "true"
L_bool_to_string_false_2:
    call instance void [lpcRuntime]LPC.File::Write(string)
    dup
    ldc.i4 13
    ldc.i4.0
    ldsfld bool _G::Puppy
    brtrue L_bool_to_string_true_3
    ldstr "false"
    br L_bool_to_string_false_4
L_bool_to_string_true_3:
    ldstr "true"
L_bool_to_string_false_4:
    call instance void [lpcRuntime]LPC.File::Format(int32, int32, string)
    call instance void [lpcRuntime]LPC.File::WriteLn()
    // line 27
    ldsflda valuetype [lpcRuntime]LPC.File _G::_output
    dup
    ldstr "The char variable is an "
    call instance void [lpcRuntime]LPC.File::Write(string)
    dup
    ldsfld uint8 _G::AnimalCookies
    call instance void [lpcRuntime]LPC.File::Write(uint8)
    call instance void [lpcRuntime]LPC.File::WriteLn()
    /* nop */
    leave L_Epilogue

L_Epilogue:
    ret

.try L_Entry to L_Epilogue finally
{
    ldsflda valuetype [lpcRuntime]LPC.File _G::_input
    call instance void [lpcRuntime]LPC.File::Close()

    ldsflda valuetype [lpcRuntime]LPC.File _G::_output
    call instance void [lpcRuntime]LPC.File::Close()

    endfinally
}

}
