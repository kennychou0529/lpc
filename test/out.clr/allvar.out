The integers are    4    5
The bytes are  212  215
The real value is      345.13    345.1250

The boolean value is true         true
The char variable is an R
